File: src/main/java/io/github/hhy/bookmark/storage/Storage.java
package io.github.hhy.bookmark.storage;

import com.intellij.openapi.project.Project;
import io.github.hhy.bookmark.element.Element;

import java.io.IOException;
import java.util.Arrays;
import java.util.List;
import java.util.Map;
import java.util.concurrent.ConcurrentHashMap;

public interface Storage {

    Map<Project, Storage> STORAGE_MAP = new ConcurrentHashMap<>();

    /**
     *
     * @return
     * @throws IOException
     */
    public List<Element> getElements() throws IOException;

    /**
     * Add
     * @param element
     */
    public void addElements(List<Element> element) throws IOException;

    default void addElement(Element element) throws IOException {
        addElements(Arrays.asList(element));
    }

    /**
     * Remove
     * @param element
     */
    public void removeElement(Element element) throws IOException;

    /**
     * storage
     * @throws IOException
     */
    public void storage() throws IOException;

    /**
     *
     * @param project
     * @return
     */
    static Storage getStorage(Project project)  {
        return STORAGE_MAP.compute(project, (k, storage) -> {
            if (storage == null) {
                storage = new LocalFileStorage(project);
            }
            return storage;
        });
    }

    static void removeStoreCache(Project project) {
        Storage remove = STORAGE_MAP.remove(project);
        try {
            remove.storage();
        } catch (IOException ignore) {

        }
    }
}

File: src/main/java/io/github/hhy/bookmark/storage/Storage.kt
package io.github.hhy.bookmark.storage

import com.intellij.openapi.project.Project
import io.github.hhy.bookmark.element.BookmarkElement
import io.github.hhy.bookmark.element.GroupElement

sealed interface Storage {

    companion object {
        @Synchronized
        fun getStorage(project: Project) = project.getService(Storage::class.java)
    }

    /**
     * 获取存储的全部书签
     */
    fun elements(): List<GroupElement>

    /**
     * 获取某个组
     */
    fun getGroup(name: String): GroupElement?

    /**
     * 添加组
     */
    fun addGroup(ele: GroupElement)

    /**
     * 删除某个组
     */
    fun removeGroup(name: String): GroupElement?

    /**
     * 重命名组
     */
    fun renameGroup(groupName: String, newGroupName: String)

    /**
     * 获取单个书签
     */
    fun getBookmark(key: String): BookmarkElement?

    /**
     * 添加书签
     */
    fun addBookmark(groupName: String, ele: BookmarkElement)

    /**
     * 删除书签
     */
    fun removeBookmark(groupName: String, key: String): BookmarkElement?

    /**
     * 存储
     */
    fun storage(): Unit
}

